File: src/#979/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var distributeCoins = function (root) {
  var res = 0
  dfs(root)
  return res

  function dfs(root) {
    if (root === null) {
      return 0
    }

    let left = dfs(root.left)
    let right = dfs(root.right)

    res += Math.abs(left) + Math.abs(right)

    return left + right + root.val - 1
  }
}

File: src/#979/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func distributeCoins(root *TreeNode) int {
	var res int = 0
	dfs(root, &res)
	return res
}

func dfs(root *TreeNode, res *int) int {
	if root == nil {
		return 0
	}

	left, right := dfs(root.Left, res), dfs(root.Right, res)
	*res += abs(left) + abs(right)
	return left + right + root.Val - 1
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
